src/abstract-interpretation/domain.ts

```diff
@@ -0,0 +1,166 @@
+}
+
+export function doIntervalsOverlap(interval1: Interval, interval2: Interval): boolean {
+	const diff1 = compareIntervals(CompareType.IgnoreInclusivity, interval1.max, interval2.min)
+	const diff2 = compareIntervals(CompareType.IgnoreInclusivity, interval2.max, interval1.min)
+
+	// If one interval ends before the other starts, they don't overlap
+	if(diff1 < 0 || diff2 < 0) {
+		return false
+	}
+	// If their end and start are equal, they only overlap if both are inclusive
+	if(diff1 === 0) {
+		return interval1.max.inclusive && interval2.min.inclusive
+	}
+	if(diff2 === 0) {
+		return interval2.max.inclusive && interval1.min.inclusive
+	}
+
+	return true
+}
+
+export function unifyDomains(domains: Domain[]): Domain {
+	const unifiedIntervals = unifyIntervals(domains.flatMap(domain => Array.from(domain.intervals)))
+	return new Domain(...unifiedIntervals)
+}
+
+export function unifyIntervals(intervals: Interval[]): Interval[] {
+	const sortedIntervals = intervals.sort(compareIntervalsByTheirMinimum)
+	if(sortedIntervals.length === 0) {
+		return []
+	}
+
+	const unifiedIntervals: Interval[] = []
+	let currentInterval = sortedIntervals[0]
+	for(const nextInterval of sortedIntervals) {
+		if(doIntervalsOverlap(currentInterval, nextInterval)) {
+			const intervalWithEarlierStart = compareIntervalsByTheirMinimum(currentInterval, nextInterval) < 0 ? currentInterval : nextInterval
+			const intervalWithLaterEnd = compareIntervalsByTheirMaximum(currentInterval, nextInterval) > 0 ? currentInterval : nextInterval
+			currentInterval = new Interval(intervalWithEarlierStart.min, intervalWithLaterEnd.max)
+		} else {
+			unifiedIntervals.push(currentInterval)
+			currentInterval = nextInterval
+		}
+	}
+	unifiedIntervals.push(currentInterval)
+	return unifiedIntervals
+}
+
+export function addDomains(domain1: Domain, domain2: Domain): Domain {
+	const intervals = new Set<Interval>()
+	for(const interval1 of domain1.intervals) {
+		for(const interval2 of domain2.intervals) {
+			intervals.add(new Interval({
+				value:     interval1.min.value + interval2.min.value,
+				inclusive: interval1.min.inclusive && interval2.min.inclusive
+			}, {
+				value:     interval1.max.value + interval2.max.value,
+				inclusive: interval1.max.inclusive && interval2.max.inclusive
+			}))
+		}
+	}
+	return new Domain(...intervals)
+}
+
+export function subtractDomains(domain1: Domain, domain2: Domain): Domain {
+	const intervals = new Set<Interval>()
+	for(const interval1 of domain1.intervals) {
+		for(const interval2 of domain2.intervals) {
+			intervals.add(new Interval({
+				value:     interval1.min.value - interval2.max.value,
+				inclusive: interval1.min.inclusive && interval2.max.inclusive
+			}, {
+				value:     interval1.max.value - interval2.min.value,
+				inclusive: interval1.max.inclusive && interval2.min.inclusive
+			}))
+		}
+	}
+	return new Domain(...intervals)
+}
+
+export function domainFromScalar(n: number): Domain {
```

Comment:
i am not too sure but conceptually i am torn if this should be (statically) part of Domain.